quickwit/quickwit-search/src/leaf.rs

```diff
@@ -424,31 +530,78 @@ pub async fn leaf_search(
     searcher_context: Arc<SearcherContext>,
     request: &SearchRequest,
     index_storage: Arc<dyn Storage>,
-    splits: &[SplitIdAndFooterOffsets],
+    splits: &mut [SplitIdAndFooterOffsets],
     doc_mapper: Arc<dyn DocMapper>,
 ) -> Result<LeafSearchResponse, SearchError> {
     let request = Arc::new(request.clone());
-    let leaf_search_single_split_futures: Vec<_> = splits
+
+    let split_filter = if request.sort_by_field.is_none() {
+        splits.sort_unstable_by(|a, b| b.split_id.cmp(&a.split_id));
+        CanSplitDoBetter::no_order()
+    } else if let Some((sort_by_field, timestamp_field)) = request
+        .sort_by_field
+        .as_deref()
+        .zip(doc_mapper.timestamp_field_name())
+    {
+        if sort_by_field == timestamp_field {
+            if request.sort_order() == SortOrder::Desc {
+                splits.sort_unstable_by_key(|b| std::cmp::Reverse(b.timestamp_end()));
+                CanSplitDoBetter::desc_timestamp()
+            } else {
+                splits.sort_unstable_by_key(|a| a.timestamp_start());
+                CanSplitDoBetter::asc_timestamp()
+            }
+        } else {
+            CanSplitDoBetter::uninformative()
+        }
+    } else {
+        CanSplitDoBetter::uninformative()
+    };
+
+    // In the future this should become `request.aggregation_request.is_some() ||
+    // request.exact_count == true`
+    let run_all_splits = true;
+
+    let mut leaf_search_single_split_futures: FuturesUnordered<_> = splits
         .iter()
         .map(|split| {
             let split = split.clone();
             let doc_mapper_clone = doc_mapper.clone();
             let index_storage_clone = index_storage.clone();
             let searcher_context_clone = searcher_context.clone();
-            let request = request.clone();
+            let mut request = (*request).clone();
+            let split_filter = split_filter.clone();
             tokio::spawn(
                 async move {
+                // TODO currently permits are not always given in order when thread pool has more
+                // than one thread. This means sorting splits is only half useful.
                 let _leaf_split_search_permit = searcher_context_clone.leaf_search_split_semaphore
                     .acquire()
                     .await
                     .expect("Failed to acquire permit. This should never happen! Please, report on https://github.com/quickwit-oss/quickwit/issues.");
+
+                if !split_filter.can_be_better(&split) {
+                    if run_all_splits {
+                        request.max_hits = 0;
```

Comment:
ah yes interesting actually.

Reply:
sadly, I don't measure any meaningful difference from that. I don't know if we have any optimization going on when `max_hits == 0` (don't download the date fastfield for instance, or somehow use less cpu)